File: rust-tests/src/bin/perf_send.rs
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, SystemTime};
use std::{env, thread};

use nethuns::sockets::{BindableNethunsSocket, NethunsSocket};
use nethuns::types::{
    NethunsCaptureDir, NethunsCaptureMode, NethunsQueue, NethunsSocketMode,
    NethunsSocketOptions,
};

#[cfg(feature = "dhat-heap")]
#[global_allocator]
static ALLOC: dhat::Alloc = dhat::Alloc;

const METER_DURATION_SECS: u64 = 10 * 60 + 1;
const METER_RATE_SECS: u64 = 10;

#[derive(Debug)]
struct Args {
    interface: String,
    batch_size: u32,
    
    numpackets: u32,
    packetsize: usize,
}


const HELP_BRIEF: &str = "\
Usage:  send [ options ]
Use --help (or -h) to see full option list and a complete description

Required options:
            [ -i <ifname> ]                     set network interface
Other options:
            [ -b <batch_sz> ]                   set batch size
            [ -s <packet_size> ]                set packet size (default 64)
            [ --numpackets <num_packets> ]      set number of packets (default 1024)
";

const PACKET_HEADER_SIZE: usize = 30;

fn main() {
    #[cfg(feature = "dhat-heap")]
    let _profiler = dhat::Profiler::new_heap();
    
    // Parse options from command line
    let args = match parse_args() {
        Ok(args) => {
            println!(
                "Test {} started with parameters: \n{:#?}",
                env::args().next().unwrap(),
                args
            );
            args
        }
        Err(e) => {
            eprintln!("Error in parsing command line options: {e}.");
            eprint!("{}", HELP_BRIEF);
            return;
        }
    };
    
    
    // Nethuns options
    let opt = NethunsSocketOptions {
        numblocks: 1,
        numpackets: args.numpackets,
        packetsize: 0,
        timeout_ms: 0,
        dir: NethunsCaptureDir::InOut,
        capture: NethunsCaptureMode::ZeroCopy,
        mode: NethunsSocketMode::RxTx,
        promisc: false,
        rxhash: false,
        tx_qdisc_bypass: true,
        ..Default::default()
    };
    
    // Generate payload
    let payload: Vec<u8> = vec![0xff; args.packetsize - PACKET_HEADER_SIZE];
    
    // Setup and fill transmission rings for each socket
    let socket = prepare_tx_socket(&args, opt).unwrap();
    
    // Define atomic variable for program termination
    let term = Arc::new(AtomicBool::new(false));
    
    // Set handler for Ctrl-C
    set_sigint_handler(term.clone());
    
    // Set timer for stopping data collection after 10 minutes
    let _ = {
        let term = term.clone();
        let stop_time = SystemTime::now()
            .checked_add(Duration::from_secs(METER_DURATION_SECS))
            .unwrap();
        thread::spawn(move || {
            if let Ok(delay) = stop_time.duration_since(SystemTime::now()) {
                thread::sleep(delay);
            }
            term.store(true, Ordering::Relaxed);
        })
    };
    
    let total = Arc::new(AtomicU64::new(0));
    let _ = {
        let total = total.clone();
        let term = term.clone();
        thread::spawn(move || {
            meter(total, term);
        })
    };
    
    let mut local_total: u64 = 0;
    
    while !term.load(Ordering::Relaxed) {
        // Prepare batch
        for _ in 0..args.batch_size {
            if socket.send(&payload).is_err() {
                break;
            }
            local_total += 1;
            
            if local_total & 0x3FF == 0 {
                // update every 1024 packets
                total.store(local_total, Ordering::Release);
            }
        }
        
        // Send batch
        socket.flush().expect("flush failed");
    }
}

/// Parses the command-line arguments and build an instance of the `Args`
/// struct.
///
/// It uses the `pico_args` crate to handle argument parsing.
///
/// # Returns
///
/// - `Ok(Args)`: If the command-line arguments are successfully parsed, a
///   Result with an Args instance containing the parsed options is returned.
/// - `Err(anyhow::Error)`: If an error occurs during argument parsing or
///   any related operations, a Result with a boxed error is returned.
fn parse_args() -> Result<Args, anyhow::Error> {
    let mut pargs = pico_args::Arguments::from_env();
    
    // Help has a higher priority and should be handled separately.
    if pargs.contains(["-h", "--help"]) {
        print!("{}", HELP_BRIEF);
        std::process::exit(0);
    }
    
    let args = Args {
        interface: pargs.value_from_str(["-i", "--interface"])?,
        batch_size: pargs.value_from_str(["-b", "--batch_size"]).unwrap_or(1),
        numpackets: pargs.value_from_str("--numpackets").unwrap_or(1024),
        packetsize: pargs.value_from_str("-s").unwrap_or(64),
    };
    
    // Check if packetsize is greater than the packet payload
    if args.packetsize <= PACKET_HEADER_SIZE {
        anyhow::bail!("Packet size must be more than {}", PACKET_HEADER_SIZE);
    }
    
    // It's up to the caller what to do with the remaining arguments.
    let remaining = pargs.finish();
    if !remaining.is_empty() {
        eprintln!("Warning: unused arguments left: {:?}.", remaining);
    }
    
    Ok(args)
}

/// Set an handler for the SIGINT signal (Ctrl-C),
/// which will notify the other threads
/// to gracefully stop their execution.
fn set_sigint_handler(term: Arc<AtomicBool>) {
    ctrlc::set_handler(move || {
        println!("Ctrl-C detected. Shutting down...");
        term.store(true, Ordering::Relaxed);
    })
    .expect("Error setting Ctrl-C handler");
}

/// Setup and fill transmission ring.
fn prepare_tx_socket(
    args: &Args,
    opt: NethunsSocketOptions,
) -> Result<NethunsSocket, anyhow::Error> {
    // Open socket
    let socket = BindableNethunsSocket::open(opt)?
        .bind(&args.interface, NethunsQueue::Any)
        .map_err(|(e, _)| e)?;
    
    Ok(socket)
}


/// Collect stats about the received packets every `METER_RATE_SECS` seconds
fn meter(total: Arc<AtomicU64>, term: Arc<AtomicBool>) {
    let mut now = SystemTime::now();
    let mut old_total: u64 = 0;
    
    while !term.load(Ordering::Relaxed) {
        // Sleep for `METER_RATE_SECS` second
        now = sleep(now);
        
        // Print number of sent packets
        let new_total = total.load(Ordering::Acquire);
        println!("{}", new_total - old_total);
        old_total = new_total;
    }
}

/// Sleep for `METER_RATE_SECS` second
fn sleep(now: SystemTime) -> SystemTime {
    let next_sys_time = now
        .checked_add(Duration::from_secs(METER_RATE_SECS))
        .expect("SystemTime::checked_add() failed");
    
    if let Ok(delay) = next_sys_time.duration_since(now) {
        thread::sleep(delay);
    }
    
    next_sys_time
}

File: cpp-tests/src/perf_send.cpp
#include <nethuns/nethuns.h>
#include <stdio.h>
#include <getopt.h>
#include <signal.h>

#include <atomic>
#include <thread>
#include <chrono>
#include <iostream>
#include <string>
#include <vector>


// help message for command line usage
const std::string help_brief = "Usage:  nethuns-send [ options ]\n" \
								"Use --help (or -h) to see full option list and a complete description.\n\n"
								"Required options: \n" \
								"\t\t\t[ -i <ifname> ] \t set network interface \n" \
								"Other options: \n" \
								"\t\t\t[ -b <batch_sz> ] \t set batch size \n" \
								"\t\t\t[ -s ] \t\t\t set packet size \n";
								

// nethuns socket
nethuns_socket_t* out = new nethuns_socket_t();
struct nethuns_socket_options netopt;
char* errbufs = new char[NETHUNS_ERRBUF_SIZE];

// configuration
uint64_t pktid = 0;
std::string interface = "";
int batch_size = 1;
unsigned int packetsize = 64;
#define PACKET_HEADER_LEN    30

// stats collection
std::atomic<uint64_t> total(0);
#define     METER_DURATION_SECS    10 * 60 + 1
#define     METER_RATE_SECS        10


// terminate application
std::atomic<bool> term(false);

// termination signal handler
void terminate(int exit_signal)
{
	(void)exit_signal;
	term.store(true, std::memory_order_relaxed);
}

void terminate_program(std::chrono::system_clock::time_point stop_timestamp) {
	std::this_thread::sleep_until(stop_timestamp);
	term.store(true, std::memory_order_relaxed);
}

void meter() {
	auto time = std::chrono::system_clock::now();
	uint64_t old_total = 0;
	
	while (!term.load(std::memory_order_relaxed)) {
		time += std::chrono::seconds(METER_RATE_SECS);
		std::this_thread::sleep_until(time);
		
		uint64_t new_total = total.load(std::memory_order_acquire);
		if (new_total < old_total) {
			// overflow detected
			exit(1);
		}
		
		std::cout << new_total - old_total << std::endl;
		old_total = new_total;
	}
}


void parse_command_line(int argc, char *argv[]);


inline std::chrono::system_clock::time_point next_meter_log() {
	return std::chrono::system_clock::now() + std::chrono::seconds(METER_RATE_SECS);
}


int main(int argc, char *argv[])
{
	parse_command_line(argc, argv);
	
	signal(SIGINT, terminate);  // register termination signal
	
	// nethuns options
	netopt = {
		.numblocks       = 1
	,   .numpackets      = 1024
	,   .packetsize      = 0
	,   .timeout_ms      = 0
	,   .dir             = nethuns_in_out
	,   .capture         = nethuns_cap_zero_copy
	,   .mode            = nethuns_socket_rx_tx
	,   .promisc         = false
	,   .rxhash          = false
	,   .tx_qdisc_bypass = true
	,   .xdp_prog        = nullptr
	,   .xdp_prog_sec    = nullptr
	,   .xsk_map_name    = nullptr
	,   .reuse_maps      = false
	,   .pin_dir         = nullptr
	};
	
	// Generate packet
	const int payload_len = packetsize - PACKET_HEADER_LEN;
	unsigned char* payload = new unsigned char[payload_len];
	memset(payload, 0xFF, packetsize - PACKET_HEADER_LEN);
	
	// Init nethuns socket
	out = nethuns_open(&netopt, errbufs);
	if (!out) {
		std::cerr << "Failed to open socket" << std::endl;
		return 1;
	}
	
	if (nethuns_bind(out, interface.c_str(), NETHUNS_ANY_QUEUE) < 0) {
		std::cerr << "Failed to bind socket to interface " << interface << std::endl;
		return 1;
	}
	
	// set up timer for stopping data collection after 10 minutes
	std::thread stop_th(
		terminate_program, 
		std::chrono::system_clock::now() + std::chrono::seconds(METER_DURATION_SECS)
	);
	
	std::thread meter_th(meter);
	
	try {
		uint64_t local_total = 0;
		
		while (!term.load(std::memory_order_relaxed)) {     
			// Prepare batch
			for (int n = 0; n < batch_size; n++) {
				if (nethuns_send(out, payload, payload_len) <= 0) {
					break;
				}
				local_total++;
				if (local_total & 0x3FF) { // update every 1024 packets
					total.store(local_total, std::memory_order_release);
				}
			}
			
			// Send batch
			nethuns_flush(out);
		}
	} catch(nethuns_exception &e) {
		if (e.sock) {
			nethuns_close(e.sock);
		}
		std::cerr << e.what() << std::endl;
		return 1;
	} catch(std::exception &e) {
		std::cerr << e.what() << std::endl;
		return 1;
	}
	
	nethuns_close(out);
	delete payload;
	return 0;
}


void parse_command_line(int argc, char *argv[])
{
	// parse options from command line
	int opt = 0;
	int optidx = 0;
	opterr = 1;     // turn on/off getopt error messages
	if (argc > 1 && argc < 10) {
		while ((opt = getopt_long(argc, argv, "hi:b:s:", NULL, &optidx)) != -1) {
			switch (opt) {
			case 'h':
				std::cout << help_brief << std::endl;
				std::exit(0);
				return;
			case 'i':
				if (optarg)
					interface = optarg;
				break;
			case 'b':
				if (optarg)
					batch_size = atoi(optarg);
				break;
			case 's':
				if (optarg)
					packetsize = atoi(optarg);
				break;
			default:
				std::cerr << "Error in parsing command line options.\n" << help_brief << std::endl;
				std::exit(-1);
				return;
			}
		}
	} else {
		std::cerr << help_brief << std::endl;
		std::exit(-1);
		return;
	}
	
	if (packetsize <= PACKET_HEADER_LEN) {
		std::cerr << "Packet size must be greater than " << PACKET_HEADER_LEN << std::endl;
		std::exit(-1);
		return;
	}

	std::cout << "\nTest " << argv[0] << " started with parameters \n"
			  << "* interface: " << interface << " \n"
			  << "* batch_size: " << batch_size << " \n"
			  << "* packetsize: " << packetsize << " \n"
			  << std::endl;
}
